File: registry/examples/registry_pong.py
#!/usr/bin/env python3
"""
Registry Pong - Server that registers with GlobalRegistry

Run GlobalRegistry first (python -m actors.registry), then run this,
then run registry_ping.py in another terminal.

Usage:
    python registry_pong.py [registry_endpoint]

THIS SOFTWARE IS OPEN SOURCE UNDER THE MIT LICENSE

Copyright 2025 Vincent Maciejewski, & M2 Tech
"""

import os
import signal
import sys

# Add actors package to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'python'))

from actors import (
    Actor, Envelope, Manager, Start, Shutdown,
    ZmqSender, ZmqReceiver, register_message,
    RegistryClient, RegistryError
)


# Define Ping message
@register_message
class Ping:
    def __init__(self, count: int = 0):
        self.count = count


# Define Pong message
@register_message
class Pong:
    def __init__(self, count: int = 0):
        self.count = count


class PongActor(Actor):
    """Receives Ping from remote, sends Pong back."""

    def on_start(self, env: Envelope) -> None:
        print("PongActor: Ready to receive pings...")

    def on_ping(self, env: Envelope) -> None:
        print(f"PongActor: Received ping {env.msg.count} from remote")
        self.reply(env, Pong(env.msg.count))


def main():
    registry_endpoint = "tcp://localhost:5555"
    if len(sys.argv) > 1:
        registry_endpoint = sys.argv[1]

    ENDPOINT = "tcp://*:5001"
    LOCAL_ENDPOINT = "tcp://localhost:5001"

    print("=== Registry Pong Process (port 5001) ===")
    print(f"Registry: {registry_endpoint}")

    # Create manager
    mgr = Manager(endpoint=LOCAL_ENDPOINT)
    handle = mgr.get_handle()

    # Create ZMQ sender/receiver for remote communication
    zmq_sender = ZmqSender(local_endpoint=LOCAL_ENDPOINT)
    zmq_receiver = ZmqReceiver(ENDPOINT, mgr, zmq_sender)

    # Create registry client and start heartbeat
    registry_client = RegistryClient("PongManager", registry_endpoint)

    try:
        registry_client.start_heartbeat()

        # Register actors with manager
        mgr.manage("zmq_receiver", zmq_receiver)
        mgr.manage("pong", PongActor())

        # Register pong actor with GlobalRegistry
        registry_client.register("pong", LOCAL_ENDPOINT)
        print(f"Registered 'pong' with GlobalRegistry at {LOCAL_ENDPOINT}")

    except RegistryError as e:
        print(f"Failed to connect to registry: {e}")
        return 1

    # Set up signal handler for clean shutdown
    def signal_handler(sig, frame):
        print("\nShutting down...")
        handle.terminate()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    # Start actors
    mgr.init()
    print("Pong process ready, 'pong' actor registered with GlobalRegistry")
    print("Press Ctrl+C to stop")

    try:
        mgr.run()
    except KeyboardInterrupt:
        pass

    # Cleanup
    registry_client.close()
    mgr.end()

    print("=== Registry Pong Process Complete ===")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: registry/examples/registry_pong.rs
/*
Registry Pong - Server that registers with GlobalRegistry

Run GlobalRegistry first (python -m actors.registry), then run this,
then run registry_ping in another terminal.

Usage:
    cd ~/actors/rust
    cargo run --example registry_pong

THIS SOFTWARE IS OPEN SOURCE UNDER THE MIT LICENSE

Copyright 2025 Vincent Maciejewski, & M2 Tech
*/

use std::sync::Arc;
use std::thread;
use std::time::Duration;

use serde::{Deserialize, Serialize};

use actors::{
    ActorContext, Manager, ThreadConfig, ManagerHandle,
    ZmqReceiver, ZmqSender, register_remote_message, define_message, handle_messages,
    RegistryClient, RegistryError,
};

// Define messages
#[derive(Serialize, Deserialize, Default)]
struct Ping {
    count: i32,
}
define_message!(Ping);

#[derive(Serialize, Deserialize, Default)]
struct Pong {
    count: i32,
}
define_message!(Pong);

/// PongActor receives Ping from remote, sends Pong back.
struct PongActor;

handle_messages!(PongActor,
    Ping => on_ping
);

impl PongActor {
    fn on_ping(&mut self, msg: &Ping, ctx: &mut ActorContext) {
        println!("PongActor: Received ping {} from remote", msg.count);
        ctx.reply(Box::new(Pong { count: msg.count }));
    }
}

fn main() {
    let registry_endpoint = std::env::args()
        .nth(1)
        .unwrap_or_else(|| "tcp://localhost:5555".to_string());

    let endpoint = "tcp://0.0.0.0:5001";
    let local_endpoint = "tcp://localhost:5001";

    println!("=== Registry Pong Process (port 5001) ===");
    println!("Registry: {}", registry_endpoint);

    // Register messages for remote serialization
    register_remote_message::<Ping>("Ping");
    register_remote_message::<Pong>("Pong");

    // Create registry client and start heartbeat
    let registry_client = RegistryClient::new("PongManager", &registry_endpoint);
    registry_client.start_heartbeat();

    // Create ZMQ sender for replies
    let zmq_sender = Arc::new(ZmqSender::new(local_endpoint));

    // Create manager and actors
    let mut mgr = Manager::new();
    let handle = mgr.get_handle();
    let pong_ref = mgr.manage("pong", Box::new(PongActor), ThreadConfig::default());

    // Create ZMQ receiver and register local actors
    let zmq_receiver = ZmqReceiver::new(endpoint, Arc::clone(&zmq_sender));
    zmq_receiver.register("pong", pong_ref);

    // Register pong actor with GlobalRegistry
    match registry_client.register("pong", local_endpoint) {
        Ok(_) => println!("Registered 'pong' with GlobalRegistry at {}", local_endpoint),
        Err(e) => {
            eprintln!("Failed to register with registry: {}", e);
            return;
        }
    }

    // Start the receiver
    let mut receiver_handle = zmq_receiver.start();

    // Set up signal handler
    let handle_clone = handle.clone();
    ctrlc::set_handler(move || {
        println!("\nShutting down...");
        handle_clone.terminate();
    }).expect("Error setting Ctrl-C handler");

    mgr.init();
    println!("Pong process ready, 'pong' actor registered with GlobalRegistry");
    println!("Press Ctrl+C to stop");

    // Wait until terminated
    mgr.run();

    // Cleanup
    receiver_handle.stop();
    mgr.end();

    println!("=== Registry Pong Process Complete ===");
}
